meteor/rsmap.py

```diff
@@ -210,7 +212,24 @@ def drop(self, labels: Any, *, inplace: bool = False) -> None | Map:
 
     def get_hkls(self) -> np.ndarray:
         # overwrite rs implt'n, return w/o modifying self -> same behavior, under testing - @tjlane
-        return self.index.to_frame().to_numpy(dtype=np.int32)
+        # this is a rather horrible thing to do and we should fix it
+        # best is to push changes upstream
+        hkl_names = ["H", "K", "L"]
+        if self.index.names == hkl_names:
+            hkls = self.index.to_frame().to_numpy(dtype=np.int32)
+        elif all(col in self.columns for col in hkl_names):
+            # we need to pull out each column as a separate DataSeries so that we don't try to
+            # create a new Map object without F, PHI
+            hkls = np.vstack([self[col].to_numpy(dtype=np.int32) for col in hkl_names]).T
+        else:
```

Comment:
In the latest version of `rs`, the index is reset somewhere along the way inside of `DataSet.to_gemmi()`, which required this additional logic.

Reply:
I'm not quite sure I follow why `rs.DataSet.get_hkls()` needs to be overwritten in the first place -- does it not work in this subclassed version? As long as the column names `["H", "K", "L"]` all exist, I would think the existing code should work for you without modifying the underlying data.

Reply:
Good question, which opens up a wider (but good!) discussion. It was definitely my hope to not overwrite these methods in mysterious ways, basically mocking their behavior. It points to something wrong in the  current implementation.

The parent implementation of `get_hkls()` does the following:
```python
self.reset_index()[["H", "K", "L"]]
```

This code performs two operations that are problematic for the current implementation of `Map`, which are:
1. insert a column with name "index", and
2. the `[[...]]`` slice creates a new instance by calling the `_constructor()` method, but the resulting `Map` will not have all the required columns. Specifically it's missing the amplitudes and phases.

Right now the implementation of `Map` enforces quite a lot of structure. My hope was to update the implementation some time soon and propose it as a feature for `rs`. It's been extremely useful for keeping the rest of the `meteor` code clean.

Reply:
Gotcha -- I could imagine changing `rs.DataSet.get_hkls()` to use the "@range_indexed" decorator which would avoid a potentially unnecessary call to `reset_index()` (which is the origin of the "index" column). 

However, the second part is a more broad impact from subclassing `rs.DataSet` (which is itself a subclass of `pd.DataFrame`), which can have some side effects here. There are a bunch of pandas operations that produce calls to `_constructor()` (for example in groupby). By only subclassing `DataSet` and not `DataSeries` as well, certain code in pandas may revert things to a `DataSet` object instead of a `Map` object. I agree that there is utility to this idea of a "limited scope and guaranteed content" sort of object, but there are likely a lot of corner cases at the moment in this implementation. This crops up here by the need to avoid certain pandas syntaxes (such as `[[col1, col2...]]`).

I'm not yet sure of a better way to achieve this cleanly at the moment, but I think we will need to think through these sorts of cases before something like `Map` could be abducted into the `rs` library

Reply:
All that said, I see why all the overloading was done here for these purposes, but I wonder if there's a more limited class definition here (that doesn't inherit all the functionality of `pandas`) that could better serve this targeted use case. That could also make this sort of thing easier to maintain in the future.

Reply:
@JBGreisman absolutely agree -- this is a great conversation, let's keep it going. I'm going to punt it over to an issue in `rs`, though, as it's bigger than this PR.